README.md
# detroit-loading-spinner

Loading spinner based on the loading spinner in the popular video game [Detroit: Become Human](https://en.wikipedia.org/wiki/Detroit:_Become_Human).

![Loading spinner screenshot](https://github.com/rafaelklaessen/detroit-loading-spinner/raw/master/screenshots/screenshot.png "Screenshot of the loading spinner")
[Demo](https://detroit-loading-spinner.netlify.com)

## Usage
```javascript
import Spinner from 'detroit-loading-spinner';

const App = () => (
  <Spinner />
);

export default App;
```

## Props
`detroit-loading-spinner` accepts the following props:

Prop name | Description | Type | Default value
--------- | ----------- | ---- | -------------
`size` | Size of the spinner | `number` | `150`
`color` | Color of the spinner | `string` | `#58728D`
`segments` | Amount of segments in spinner | `number` | `4`
`segmentMax` | Maximum size (in percents) of one spinner segment | `number` | `65`
`padding` | Space between the spinner segments | `number` | `2`
`cornerRadius` | Corner radius of the spinner segments | `number` | `20`

Undocumented props are passed on the root element.


## License
MIT


CHANGELOG.md
# Version 1.0.0
Released 2018-08-04

- Re-release of initial release

# Version 1.0.0-rc.0
Released 2018-08-04

- Initial release


src/Arc.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as d3 from 'd3';
import { arc as d3Arc } from 'd3-shape';

export default class Arc extends Component {
  static propTypes = {
    size: PropTypes.number.isRequired,
    startAngle: PropTypes.number.isRequired,
    endAngle: PropTypes.number.isRequired,
    cornerRadius: PropTypes.number
  };

  static defaultProps = {
    cornerRadius: 20
  };

  id = Math.random();

  arc = d3Arc()
    .innerRadius(this.props.size / 2.75)
    .outerRadius(this.props.size / 2)
    .startAngle(this.props.startAngle)
    .endAngle(this.props.endAngle)
    .cornerRadius(this.props.cornerRadius);

  componentDidUpdate = prevProps => {
    d3.select(document.getElementById(this.id))
      .transition()
      .ease(d3.easeExpOut)
      .duration(1000)
      .attrTween('d', () => step =>
        this.arc
          .startAngle(
            this.calculateNextStartAngle(
              prevProps.startAngle,
              this.props.startAngle,
              step
            )
          )
          .endAngle(
            this.calculateNextEndAngle(
              prevProps.endAngle,
              this.props.endAngle,
              step
            )
          )()
      );
  };

  calculateNextStartAngle = (oldStartAngle, newStartAngle, step) =>
    oldStartAngle + step * (newStartAngle - oldStartAngle);

  calculateNextEndAngle = (oldEndAngle, newEndAngle, step) =>
    oldEndAngle + step * (newEndAngle - oldEndAngle);

  render() {
    const arc = this.arc();
    const { size, startAngle, endAngle, cornerRadius, ...props } = this.props;
    return (
      <g>
        <path id={this.id} d={arc} {...props} />
      </g>
    );
  }
}


src/Pie.js
import React from 'react';
import PropTypes from 'prop-types';
import { pie as d3Pie } from 'd3-shape';
import Arc from './Arc';
import { ArcTypes } from './index';

const Pie = ({ size, color, cornerRadius, data, opacity }) => {
  const pie = d3Pie()
    .value(d => d.value)
    .sort(() => 0);
  return (
    <g transform={`translate(${size / 2}, ${size / 2})`}>
      {pie(data).map((arc, i) =>
        <Arc
          size={size}
          cornerRadius={cornerRadius}
          startAngle={arc.startAngle}
          endAngle={arc.endAngle}
          fill={color}
          fillOpacity={arc.data.type === ArcTypes.FILLED ? opacity : 0}
          key={i}
        />
      )}
    </g>
  );
};

Pie.propTypes = {
  size: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
  cornerRadius: PropTypes.number.isRequired,
  data: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.oneOf(['filled', 'blank']),
    value: PropTypes.number
  })).isRequired,
  opacity: PropTypes.number.isRequired
};

export default Pie;


src/index.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Pie from './Pie';

export const ArcTypes = {
  FILLED: 'filled',
  BLANK: 'blank'
};

export default class Spinner extends Component {
  static propTypes = {
    size: PropTypes.number,
    color: PropTypes.string,
    segments: PropTypes.number,
    segmentMax: PropTypes.number,
    padding: PropTypes.number,
    cornerRadius: PropTypes.number
  };

  static defaultProps = {
    size: 150,
    color: '#58728D',
    segments: 4,
    segmentMax: 65,
    padding: 2,
    cornerRadius: 20
  };

  constructor(props) {
    super(props);
    this.state = {
      pieOneData: this.generatePercentages(),
      pieTwoData: this.generatePercentages()
    };
  }

  intervalId = null;

  componentDidMount = () => {
    this.intervalId = setInterval(this.updateProgress, 1000);
  };

  componentWillUnmount = () => clearInterval(this.intervalId);

  updateProgress = () =>
    this.setState({
      pieOneData: this.generatePercentages(),
      pieTwoData: this.generatePercentages()
    });

  generatePercentages = () => {
    const percentagesWithoutPadding = this.generatePercentagesWithoutPadding();
    const startsWithPadding = Math.random() >= .5;
    return this.addPaddingToPercentages(
      percentagesWithoutPadding,
      startsWithPadding
    );
  };

  generatePercentagesWithoutPadding = () => {
    const segments = this.props.segments;
    let segmentMax = this.props.segmentMax;
    // https://stackoverflow.com/a/39630127
    const max = 100;
    const tempResults = [];
    let remaining = max;
    let finalResults = [];

     // create a series of random numbers and push them into an array
    for (let i = 1; i <= segments; i++) {
      let r = Math.random() * segmentMax;
      if (i === segments) {
        // the final segment is just what's left after the other randoms are
        // added up
        r = remaining;
      }
      tempResults.push(r);
      // subtract them from the total
      remaining -= r;
      // no segment can be larger than what's remaining
      segmentMax = remaining;
    }

    // randomly shuffle the array into a new array
    while (tempResults.length > 0) {
      const index = Math.floor(Math.random() * tempResults.length);
      finalResults = finalResults.concat(tempResults.splice(index, 1));
    }
    return finalResults;
  };

  addPaddingToPercentages = (percentages, startsWithPadding) => {
    let availablePadding = 0;
    const targetPadding = this.props.padding;
    const withPadding = percentages.map((percentage) => {
      if (percentage > targetPadding) {
        availablePadding += targetPadding;
        return percentage - targetPadding;
      }
      availablePadding += percentage;
      return 0;
    });

    const padding = availablePadding / 8;

    const percentagesAndPaddings = [];

    percentagesAndPaddings.push({
      type: ArcTypes.BLANK,
      value: startsWithPadding ? padding : 0
    });

    withPadding.forEach((percentage) => {
      percentagesAndPaddings.push({
        type: ArcTypes.FILLED,
        value: percentage
      });
      percentagesAndPaddings.push({
        type: ArcTypes.BLANK,
        value: padding
      });
    });

    if (!startsWithPadding) {
      percentagesAndPaddings[percentagesAndPaddings.length - 1].value = 0;
    }

    return percentagesAndPaddings;
  };

  render() {
    const { pieOneData, pieTwoData } = this.state;
    const {
      size,
      color,
      segments,
      segmentMax,
      padding,
      cornerRadius,
      ...props
    } = this.props;
    return (
      <svg width={size} height={size} {...props}>
        <Pie
          size={size}
          cornerRadius={cornerRadius}
          data={pieOneData}
          color={color}
          opacity={.3}
        />
        <Pie
          size={size}
          cornerRadius={cornerRadius}
          data={pieTwoData}
          color={color}
          opacity={.7}
        />
      </svg>
    );
  }
}
